tests/dualtor/test_tor_ecn.py

```diff
@@ -0,0 +1,344 @@
+
+    inner_dscp = random.choice(range(0, 33))
+    inner_ttl = random.choice(range(3, 65))
+    inner_ecn = random.choice(range(0,3))
+
+    inner_packet = testutils.simple_ip_packet(
+        ip_src="1.1.1.1",
+        ip_dst=server_ipv4,
+        ip_dscp=inner_dscp,
+        ip_ttl=inner_ttl,
+        ip_ecn=inner_ecn
+    )[IP]
+    packet = testutils.simple_ipv4ip_packet(
+        eth_dst=tor.facts["router_mac"],
+        eth_src=ptfadapter.dataplane.get_mac(0, 0),
+        ip_src=peer_ipv4_address,
+        ip_dst=tor_ipv4_address,
+        ip_dscp=inner_dscp,
+        ip_ttl=255,
+        ip_ecn=inner_ecn,
+        inner_frame=inner_packet
+    )
+    logging.info("the encapsulated packet to send:\n%s", dump_scapy_packet_show_output(packet))
+
+    return packet
+
+@pytest.fixture(scope="function")
+def build_non_encapsulated_ip_packet(
+    rand_selected_interface, 
+    ptfadapter, 
+    rand_selected_dut, 
+    tunnel_traffic_monitor
+):
+    """
+    Build the regular (non encapsulated) packet to be sent from T1 to ToR.
+    """
+    tor = rand_selected_dut
+    _, server_ips = rand_selected_interface
+    server_ipv4 = server_ips["server_ipv4"].split("/")[0]
+    config_facts = tor.get_running_config_facts()
+    try:
+        peer_ipv4_address = [dut_name["address_ipv4"] for dut_name in config_facts["PEER_SWITCH"].values()][0]
+    except IndexError:
+        raise ValueError("Failed to get peer ToR address from CONFIG_DB")
+
+    tor_ipv4_address = [addr for addr in config_facts["LOOPBACK_INTERFACE"]["Loopback0"]
+                        if is_ipv4_address(addr.split("/")[0])][0]
+    tor_ipv4_address = tor_ipv4_address.split("/")[0]
+
+    dscp = random.choice(range(0, 33))
+    ttl = random.choice(range(3, 65))
+    ecn = random.choice(range(0,3))
+
+    packet = testutils.simple_ip_packet(
+        eth_dst=tor.facts["router_mac"],
+        eth_src=ptfadapter.dataplane.get_mac(0, 0),
+        ip_src="1.1.1.1",
+        ip_dst=server_ipv4,
+        ip_dscp=dscp,
+        ip_ecn=ecn,
+        ip_ttl=ttl
+     )
+    logging.info("the regular IP packet to send:\n%s", dump_scapy_packet_show_output(packet))
+
+    return packet
+
+def get_ptf_server_intf_index(
+    tor, 
+    tbinfo, 
+    iface
+):
+    """
+    Get the index of ptf ToR-facing interface on ptf.
+    """
+    mg_facts = tor.get_extended_minigraph_facts(tbinfo)
+
+    return mg_facts["minigraph_ptf_indices"][iface]
+
+def build_expected_packet_to_server(
+    encapsulated_packet
+):
+    """
+    Build packet expected to be received by server from the tunnel packet.
+    """
+    inner_packet = encapsulated_packet[IP].payload[IP].copy()
+    # use dummy mac address that will be ignored in mask
+    inner_packet = Ether(src="aa:bb:cc:dd:ee:ff", dst="aa:bb:cc:dd:ee:ff") / inner_packet
+    exp_pkt = mask.Mask(inner_packet)
+    exp_pkt.set_do_not_care_scapy(Ether, "dst")
+    exp_pkt.set_do_not_care_scapy(Ether, "src")
+    exp_pkt.set_do_not_care_scapy(IP, "tos")
```

Comment:
Why are we ignoring "tos" field? should this be the same?

Reply:
I am checking DSCP as well as ECN in individual test cases although ToS should be the same.